src/v/raft/heartbeats.cc

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "raft/heartbeats.h"
+
+#include "raft/types.h"
+#include "serde/serde.h"
+
+#include <cstdint>
+
+namespace raft {
+void heartbeat_request_v2::add(const group_heartbeat& hb) {
+    if (hb.data) {
+        _full_heartbeats.push_back(
+          full_heartbeat{.group = hb.group, .data = *hb.data});
+    } else {
+        _lw_heartbeats.add(hb.group);
+    }
+}
+
+heartbeat_request_v2 heartbeat_request_v2::copy() const {
+    heartbeat_request_v2 ret;
+    ret._source_node = _source_node;
+    ret._target_node = _target_node;
+    ret._lw_heartbeats = _lw_heartbeats.copy();
+    ret._full_heartbeats.reserve(_full_heartbeats.size());
+
+    std::copy(
+      _full_heartbeats.begin(),
+      _full_heartbeats.end(),
+      std::back_inserter(ret._full_heartbeats));
+    ret._lw_cnt = _lw_cnt;
+
+    return ret;
+}
+
+ss::future<> heartbeat_request_v2::serde_async_write(iobuf& out) {
+    using serde::write;
+    using serde::write_async;
+    write(out, _source_node);
+    write(out, _target_node);
+    write(out, _lw_cnt);
+
+    co_await write_async(out, std::move(_lw_heartbeats));
+    co_await write_async(out, std::move(_full_heartbeats));
```

Comment:
it looks like there is no async serialization neither for `iobuf` nor `chunked_fifo` i will add an explicit yield in between, it would be great to provide async serialization for collections at least